### ISSUE TITLE
Unable to quantize weights of pytorch jit scripted model with an LSTM layer unless quant_pre_process is run

### ISSUE THREAD

**Issue by dhruvbird** [microsoft/onnxruntime#18671]
### Describe the issue

If I have a PyTorch model with an LSTM layer `quantize_dynamic()` is able to reduce the model size if fed the original python model instance, but is unable to reduce the model size if fed in a jit scripted version of the same model.

### To reproduce

```
import torch
from torch import nn
import os
import onnx


class SimpleLSTMModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.embed = nn.Embedding(32, 128, padding_idx=0)
        self.inner = nn.LSTM(128, 256, 1, batch_first=True)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = self.embed(x)
        x, _ = self.inner(x)
        return x


m = SimpleLSTMModel()
scripted = torch.jit.script(m)


def quantize_and_model_size(model_path, model):
    with open(model_path, "wb") as fp:
        torch.onnx.export(
            model,
            (torch.tensor([[2, 3, 1, 9], [6, 1, 0, 7]]),),
            fp,
            input_names=["x"],
            dynamic_axes={"x": {1: "seq_length"}},
        )

    quantize_dynamic(model_path, "/tmp/q.onnx")
    file_size_orig = os.stat(model_path).st_size
    file_size_quant = os.stat("/tmp/q.onnx").st_size

    print(
        f"Original model size: {file_size_orig}, Quantized model size: {file_size_quant}"
    )


quantize_and_model_size("/tmp/simple_lstm.onnx", m)
quantize_and_model_size("/tmp/simple_lstm.onnx", scripted)
```

Output:

```
Original model size: 1599244, Quantized model size: 408875
Original model size: 1599160, Quantized model size: 1588135
```

However, when it's preprocessed, the it behaves like the outcome on the python model.

```
import torch
from torch import nn
import os
import onnx


class SimpleLSTMModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.embed = nn.Embedding(32, 128, padding_idx=0)
        self.inner = nn.LSTM(128, 256, 1, batch_first=True)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = self.embed(x)
        x, _ = self.inner(x)
        return x


m = SimpleLSTMModel()
scripted = torch.jit.script(m)


def quantize_and_model_size(model_path, model):
    with open(model_path, "wb") as fp:
        torch.onnx.export(
            model,
            (torch.tensor([[2, 3, 1, 9], [6, 1, 0, 7]]),),
            fp,
            input_names=["x"],
            dynamic_axes={"x": {1: "seq_length"}},
        )

    quant_pre_process(model_path, "/tmp/pp.onnx")
    quantize_dynamic("/tmp/pp.onnx", "/tmp/q.onnx")
    file_size_orig = os.stat(model_path).st_size
    file_size_quant = os.stat("/tmp/q.onnx").st_size

    print(
        f"Original model size: {file_size_orig}, Quantized model size: {file_size_quant}"
    )


quantize_and_model_size("/tmp/simple_lstm.onnx", m)
quantize_and_model_size("/tmp/simple_lstm.onnx", scripted)
```

Output:

```
Original model size: 1599244, Quantized model size: 408966
Original model size: 1599160, Quantized model size: 408912
```

### Urgency

Everything is always urgent :)

### Platform

Mac

### OS Version

n/a

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

version = "1.15.0" git_version = "b86cc54efce19530fb953e4b21f57e6b3888534c"

### ONNX Runtime API

Python

### Architecture

ARM64

### Execution Provider

Default CPU

**Comment by yufenglee** [comment]
It is highly recommend to run the preprocess before quantization: https://onnxruntime.ai/docs/performance/model-optimizations/quantization.html#pre-processing. The pre-processing can infer data type and fold constant subgraph, which are important to quantization.

**Comment by dhruvbird** [comment]
Fair point - I'll remember to do this - it's just that sometimes the preprocess step fails if there is a for-loop in the model and it's been challenging to deal with. I'll open another issue for that one in case I continue to be blocked by it.
